react/hooks/useFadeEffect.ts

```diff
@@ -0,0 +1,141 @@
+import {
+  useCallback,
+  useEffect,
+  useLayoutEffect,
+  useReducer,
+  useRef,
+} from 'react'
+
+const TRANSITION_DURATION = 1000
+
+interface State {
+  isTransitioning: boolean
+  shouldBeVisible: boolean
+}
+
+type Action = { type: 'start'; shouldBeVisible: boolean } | { type: 'finish' }
+
+function reducer(state: State, action: Action) {
+  switch (action.type) {
+    case 'start':
+      return { isTransitioning: true, shouldBeVisible: action.shouldBeVisible }
+    case 'finish':
+      return { isTransitioning: false, shouldBeVisible: state.shouldBeVisible }
+    default:
+      return state
+  }
+}
+
+export const useFadeEffect = (visible: boolean) => {
+  const elementRef = useRef<HTMLElement | null>(null)
+
+  const [state, dispatch] = useReducer(reducer, {
+    isTransitioning: false,
+    shouldBeVisible: false,
+  })
+
+  const { isTransitioning, shouldBeVisible } = state
+
+  const timeoutIdRef = useRef<ReturnType<typeof setTimeout> | null>(null)
+  const rafIdRef = useRef<number | null>(null)
```

Comment:
question: what does "raf" means?

Reply:
short for "requestAnimationFrame" 😅

Reply:
wow, could you explicit this in a comment, please? Because even after some googling, I didn't found a relevant result lol

Reply:
do you think typing it as the following could help?

```ts
const rafIdRef = useRef<ReturnType<typeof requestAnimationFrame> | null>(null)
```

Reply:
Looks even better to me!

Reply:
updated in 908bbf6 :-)